### ISSUE TITLE
'Changes' display GB2312 contents incorrectly

### ISSUE THREAD

**Issue by weyllor** [ForkIssues/TrackerWin#173]
My files is encoded by GB2312, with Chinese characters.
And the content changes are not display correctly.
[gb2312.txt](https://github.com/ForkIssues/TrackerWin/files/2904395/gb2312.txt)
![image](https://user-images.githubusercontent.com/274591/53411911-cde20c80-3a02-11e9-846b-0f0f0bfd456a.png)

BTW, it's OK if the file is converted to UTF-8.
![image](https://user-images.githubusercontent.com/274591/53411954-f0742580-3a02-11e9-81a3-93ad970d14e9.png)

**Comment by gageas** [comment]
Hello.
As far as I know, most git tool have same behavior, SourceTree, GitHub Desktop, etc.
Please consider to set **working-tree-encoding** attribute in `.gitattributes` file.
[https://git-scm.com/docs/gitattributes](https://git-scm.com/docs/gitattributes#_effects)

It works in my case, not GB2312 but  CP932 a.k.a. Shift_JIS.

**Comment by weyllor** [comment]
> Hello.
> As far as I know, most git tool have same behavior, SourceTree, GitHub Desktop, etc.
> Please consider to set **working-tree-encoding** attribute in `.gitattributes` file.
> [https://git-scm.com/docs/gitattributes](https://git-scm.com/docs/gitattributes#_effects)
> 
> It works in my case, not GB2312 but CP932 a.k.a. Shift_JIS.

Thanks for you reply @gageas .
Maybe we can detect the charset of a file, and show the file content and diff content better.
Sometimes a project use several charsets for different files, a project-wide attribute may not handle this case.

**Comment by gageas** [comment]
> Sometimes a project use several charsets for different files, a project-wide attribute may not handle this case.

You can specify attribute for file extension, directory path, or single filename.
In my project, *.c / *.h files are written in Shift_JIS. And *.md files are written in UTF8.

Auto detection of charset is not trivial task and always not successful.

**Comment by hitzhangjie** [comment]
Are the developers ready to fix this bug? Has come across the same problem.

**Comment by hitzhangjie** [comment]
@weyllor Hi，have you solved this problem？

**Comment by hitzhangjie** [comment]
@DanPristupov Do we have any devopers have an eye on this ?

**Comment by DanPristupov** [comment]
No, I did not have a time for that yet. Please consider using the unicode encoding.

**Comment by weyllor** [comment]
Maybe I can contribute some code for this issue, if I can get some related code.

**Comment by BianChengNan** [comment]
same on me. I think tortoisegit solves this kind problem well. hope it can be supported in fork too.

**Comment by fc-reus** [comment]
SourceTree has an option to set the global encoding, so practically it should be possible to fix this in some way.
Nevertheless in SourceTree it's not ideal that this is a global setting and not adjustable per project.

**Comment by fc-reus** [comment]
It would be great if at least iso-8859-1 would be supported like in git, it supports utf-8 and iso-8859-1 by default, some other encodings need special configuration.

> Git recognizes files encoded in ASCII or one of its supersets (e.g. UTF-8, ISO-8859-1, …​) as text files. Files encoded in certain other encodings (e.g. UTF-16) are interpreted as binary and consequently built-in Git text processing tools (e.g. git diff) as well as most Git web front ends do not visualize the contents of these files by default.

And GitHub by the way seems to support it in the source blob (but not in the raw view, which is interpeted in a wrong way because the browser can't detect the encoding for the raw view which is a actually a web view).

**Comment by anthroid** [comment]
I've found a workaround for this in my particular use case (hopefully the same for others). Commenting here since it looks like all other related issues reference this one.

Since Fork (implicitly) supports the `diff=` attribute in `.gitattributes` and the corresponding attribute definition in `.gitconfig`, it is possible to pass the output to another command to convert the encoding on the fly before it is interpreted and displayed in the GUI. This is normally used for binary files in order to get some kind of readable plain text output.

In my case, I have a large code base managed by a proprietary IDE which insists on saving with `cp1252` encoding (and can't be changed by the user), and frequently contains special accented characters like `éêëïô`. These are all rendered as `�` by Fork and similarly broken as `<E9>` by `git diff` on the command line like the screenshot in this issue.

The solution is to pipe the `git` output through `iconv` before it is seen by Fork (or the command line output). If you know the encoding in advance, you can set it for each file extension and configure the command as follows:

### .gitattributes or .git/info/attributes
Replace `ext` with your file extension. The `diff=` part can be any name.
```ini
*.ext text diff=ext2utf
```

### ~/.gitconfig or .git/config
Replace `"ext2utf"` with the name you used in `.gitattributes`.
Replace `-f cp1252` with your file's original encoding. Check `iconv -l` for a list of encodings on your machine.
```ini
[diff "ext2utf"]
        textconv = iconv -f cp1252 -t utf-8
```
Alternatively, using `git config`
```
git config [--global|--local] diff.ext2utf.textconv 'iconv -f cp1252 -t utf-8'
```

@DanPristupov Would it be possible for Fork to use this mechanism? Possibly even combined with `file -I` to automatically interpret the original encoding?
